fdbserver/include/fdbserver/MovingWindow.h

```diff
@@ -27,8 +27,16 @@
 #include "flow/Deque.h"
 #include "fdbserver/Knobs.h"
 
-// Perfomed as the rolling window to calculate average change rates in the past <interval>
-// e.g., we may use it in "MovingData" Trace to show average moving bytes rate by DD.
+// Smoother will try to modify the sum of all samples and return an estimtaed value, while
+// MovingWindow will just sum up all the samples, and getAverage() will return the average
+// change rate in the past <timeWindow>.
+// The motivation of MovingWindow comes from "MovingData" trace, where we want to get the
+// average bytes moved by DD in the past DD_TRACE_MOVE_BYTES_AVERAGE_INTERVAL. This data could be
+// precise and clear to show how DD works recently.
+// As for Smoother, however, it doesn't have a method to calculate average change rate, and it will
+// take all of the samples into account, while we don't really need it for "MovingData". Also, saying there
+// is no DD workload recently, we could just return 0 with MovingWindow and Smoother will return
+// a value closed to 0, where the former method is more obvious and clear to understand DD workloads.
```

Comment:
I'm not sure this is an accurate characterization of `Smoother`. The smoother uses exponential smoothing to more heavily weight recent samples, and for the motivation described here in DD it sounds like it would be a reasonably good fit for the Smoother. The smoother also does provide both a rate and a total.

A moving average equally weights all of the samples across the window, which has some slightly different properties that may be useful in some circumstances. It could also be a good fit for the DD case, but at the moment it's not clear to me why it would be preferable to the Smoother.

Reply:
I think the main difference is that `MovingWindow` will show the actual numbers while `Smoother` is going to show the trend. To specify, one difference is "accurate", we would get to know the exact bytes move rate by DD with MovingWindow, while Smoother would provide a smoothed rate. Another is how we measure the time window. MovingWindow will ignore the samples before the time window, while Smoother will attach less weight to them.
In the above DD case:
```
elapsed = 5, eFoldingTime = 120,
estimate = estimate + (total - estimate) * (1 - exp(-elapsed / eFoldingTime))  = estimate + (total - estimate) * 0.04;
```
Saying there is only one sample `delta = 100 bytes`, MovingWindow would plot a histogram which remains `total - previous = 100`(didn't include `/ interval` calculation for simplicity) through 120 seconds and goes down to `0`. Meanwhile, `smoothRate` will plot a line chart(`total - estimate`, didn't include `/eFoldingTime`). The data points of two graphs show like this:
```
MovingWindow:
100, 100, ..., 100(120 seconds), 0(125 seconds), 0...
Smoother:
96 92.16 88.4736 84.9347 81.5373 78.2758 75.1447 72.139 69.2534 66.4833 63.8239 61.271 58.8201 56.4673 54.2086 52.0403 49.9587 47.9603 46.0419 44.2002 42.4322 40.7349 39.1055 37.5413(120 seconds), 36.0397(125 seconds)...
```
I think it's probably more clear to have a histogram with MovingWindow to specify there is a single DD movement operation which moved 100 bytes in the past 120 seconds. With Smoother, however, we don't know how many bytes it moved, and we don't know when the time window ends.
After the time window, MovingWindow will show `0`, while Smoother will follow the same rhythm and goes to `0`(never reached) more and more slowly. It might be clear to have a `0` after 120 seconds since we would exactly know there is no DD operations for the last `eFoldingTime`. I guess we can also get to know that by telling Smoother graph keeps going down. But it will need some background and context to know that.
Overall, they do behave similar, and I thought `MovingWindow` might be a better fit.

Reply:
And to clarify, I don't think it's wrong to want to use a different smoothing function here. I just want to make sure that the justification is clear in the comments here so that the rationale is both well-understood and can be applied appropriately for future use-cases of either.

Reply:
Perhaps a nice way to address this (but maybe not needed at the moment) is to make an `ISmoother` interface and then have an `ExponentialSmoother` and a `MovingAverage` smoother implementation. Then the documentation for each could describe the characteristics of that smoother in a pretty simple manner and then the `ISmoother` interface could describe the general properties of a generic smoother.

Reply:
> To specify, one difference is "accurate", we would get to know the exact bytes move rate by DD with MovingWindow
>
Ah, the word "accurate" might not be precise, what I wanted to specify is that `MovingWindow` could return the actual moved bytes rate, which is more favorable in the DD case.

Reply:
Yes, I think both of them make sense in its own use cases and a generic `ISmoother` interface might be the best idea to unit all of two smooth ways.
I just submitted a new commit and hopefully it explains more clearly.